lib/core.js

```diff
@@ -323,6 +323,7 @@ declare class String {
     valueOf(): string;
     toString(): string;
     length: number;
+    [key: number]: string;
```

Comment:
Should this instead be `+[key: number]: string | void`?

I mean, it's definitely covariant, but should the `| void` be added?

```js
let a = ''
a[10].toString()
```

Reply:
That was also my gut reaction but I chose to omit the `| void` for compatibility with the [`Array` indexer property](https://github.com/facebook/flow/blob/master/lib/core.js#L277).

None the less, I agree with you. This definitely feels like a bug in the `Array` lib def _([example](https://flow.org/try/#0PTAEAEDMBsHsHcBQiDGsB2BnALqAhgE4F4CeAXKAIJGkA86ArgLYBGApgQHygC8oA2gEYANKABMogMwBdVBhygAbnmgM2FZara98NEvwC0g2UA))_.

Reply:
I feel like you should copy [the definition for $ReadOnlyArray](https://github.com/facebook/flow/blob/master/lib/core.js#L237-L238) and make both the `length` and indexer properties covariant.

Reply:
That makes sense semantically although I don't think that makes the `string` indexer property any more or less sound.

**Rational**

Writing into index `[0]` on `Array<*>` with a `number` and then index `[1]` with a `string` produces `Array<number>` and then `Array<number | string>` respectively. The assignment to the indexer property will change the type of it's value (`T`). In this case covariance is beneficial because it guarantees the value of `T` does not change at "compile time". 

On the other hand, `string` is monomorphic and immutable at the language level so we do not need to worry about the type of the indexer property value from changing. In this case, making the indexer property covariant only makes the type less flexible.

**tldr;**

In my opinion, keeping the indexer property bivariant provides greater flexibility without sacrificing soundness. 

In any case, [I could be missing something](https://github.com/facebook/flow/pull/4713#issuecomment-324923277). I am happy to amend this PR if the bivariant indexer property is unsound. 😃

Reply:
Now that I think about it, I'm wondering if this isn't part of string because of Unicode issues. Subscripting directly into a string isn't safe because you may get part of a multi-byte Unicode character. If you want a character array you are supposed to get the string's iterator, which will let you iterate over the Unicode characters.

Reply:
Ahhh that's a great point. Funny enough, you just made me find a bug in an iterator library I'm working on! Still, that feels better suited lint warning than a type error.

For example you may want to write the bytes of a string into some arbitrary destination such as Uint8Array.

Reply:
Hmm, yeah, but in some sense the right return type for string subscript isn't string, but rather a code point or something like that...

Reply:
We need more types!

Reply:
https://doc.rust-lang.org/std/primitive.char.html

Reply:
Hmm after giving it some thought I still feel like the return type for string subscript is string. 
`'🍕'[0].constructor` is `String`. To me, from a type perspective that is safe. You can call `.toUpperCase()` without a run-time error. Your 100% right that it is potentially incorrect but I feel like that would be a logic error at the application level (at least in JavaScript).